app/uk/gov/hmrc/cdsreimbursementclaimfrontend/controllers/upload/FileUpload.scala

```diff
@@ -0,0 +1,147 @@
+import uk.gov.hmrc.cdsreimbursementclaimfrontend.models.upscan.{ScheduledDocument, SupportingEvidence, UploadReference, UpscanUpload}
+import uk.gov.hmrc.cdsreimbursementclaimfrontend.controllers.supportingevidence.{routes => evidenceRoutes}
+import uk.gov.hmrc.cdsreimbursementclaimfrontend.models.JourneyStatus.FillingOutClaim
+import uk.gov.hmrc.cdsreimbursementclaimfrontend.models.upscan.UpscanCallBack.UpscanSuccess
+
+trait FileUpload[A] {
+
+  val key: String
+
+  val maxUploads: Int
+
+  def hasReachedUploadThreshold(maybeAnswer: Option[A]): Boolean
+
+  def hasReference(maybeAnswer: Option[A], uploadReference: UploadReference): Boolean
+
+  def add(
+    upload: UpscanUpload,
+    callback: UpscanSuccess,
+    maybeAnswer: Option[A],
+    claim: FillingOutClaim
+  ): FillingOutClaim
+
+  def reviewPage: Call
+
+  def uploadErrorPage: Call
+
+  def handleUploadCallback(uploadReference: UploadReference): Call
+
+  def scanSuccessPage(uploadReference: UploadReference): Call
+
+  def scanErrorPage: Call
+}
+
+@Singleton
+class FileUploadServices @Inject() (config: Configuration) {
+
+  implicit object SupportingEvidenceUpload extends FileUpload[SupportingEvidencesAnswer] {
+
+    val key: String     = "supporting-evidence"
+    val maxUploads: Int = config.readMaxUploadsValue(key)
+
+    def hasReachedUploadThreshold(maybeAnswer: Option[SupportingEvidencesAnswer]): Boolean =
+      maybeAnswer.exists(_.length >= maxUploads)
+
+    def hasReference(maybeAnswer: Option[SupportingEvidencesAnswer], uploadReference: UploadReference): Boolean =
+      maybeAnswer.exists(_.exists(_.uploadReference === uploadReference))
+
+    def add(
+      upload: UpscanUpload,
+      callback: UpscanSuccess,
+      maybeAnswer: Option[SupportingEvidencesAnswer],
+      fillingOutClaim: FillingOutClaim
+    ): FillingOutClaim = {
+
+      val newEvidence = SupportingEvidence(
+        upload.uploadReference,
+        upload.upscanUploadMeta,
+        upload.uploadedOn,
+        callback,
+        callback.fileName,
+        None
+      )
+
+      val evidences = maybeAnswer.map(_ :+ newEvidence) orElse Some(SupportingEvidencesAnswer(newEvidence))
+
+      FillingOutClaim.of(fillingOutClaim)(_.copy(supportingEvidencesAnswer = evidences))
+    }
+
+    def uploadErrorPage: Call = evidenceRoutes.SupportingEvidenceController.handleUpscanErrorRedirect()
+
+    def handleUploadCallback(uploadReference: UploadReference): Call =
+      evidenceRoutes.SupportingEvidenceController.scanProgress(uploadReference)
+
+    def scanErrorPage: Call = evidenceRoutes.SupportingEvidenceController.handleUpscanCallBackFailures()
+
+    def scanSuccessPage(uploadReference: UploadReference): Call =
+      evidenceRoutes.SupportingEvidenceController.chooseSupportingEvidenceDocumentType(uploadReference)
+
+    def reviewPage: Call = evidenceRoutes.SupportingEvidenceController.checkYourAnswers()
+  }
+
+  implicit object ScheduledDocumentUpload extends FileUpload[ScheduledDocument] {
```

Comment:
`ScheduleDocument` should be called `ScheduleDocumentAnswer` to make it consistent with the supporting-evidence upload implementation